src/backend/tcop/idle_resource_cleaner.c

```
@@ -0,0 +1,134 @@
+{
+	DisconnectAndDestroyUnusedGangs();
+}
+
+int IdleSessionTimeout = 0;
+
+int (*get_idle_session_timeout_hook)(void) = NULL;
+void (*idle_session_timeout_action_hook)(void) = NULL;
+
+
+/*
+ * Given two timeout values (e.g. to be passed to enable_sig_alarm), returns
+ * the earlier one, ignoring invalid (non-positive) values. If both are invalid,
+ * returns a non-positive number.
+ */
+static int MinTimeout(int x, int y)
+{
+	if (y <= 0) return x;
+	if (x <= 0) return y;
+
+	return Min(x, y);
+}
+
+/*
+ * We want to check to see if our session goes "idle" (nobody sending us work to
+ * do). We decide this is true if after waiting a while, we don't get a message
+ * from the client.
+ * We can then free resources (right now, just the gangs on the segDBs).
+ *
+ * A bit ugly:  We share the sig alarm timer with the deadlock detection.
+ * We know which it is (deadlock detection needs to run or idle
+ * session resource release) based on the DoingCommandRead flag.
+ *
+ * Note we don't need to worry about the statement timeout timer
+ * because it can't be running when we are idle.
+ *
+ * We want the time value to be long enough so we don't free gangs prematurely.
+ * This means giving the end user enough time to type in the next SQL statement
+ *
+ *
+ * GPDB_93_MERGE_FIXME: replace the enable_sig_alarm() calls with the
+ * new functionality provided in timeout.c from PG 9.3.
+ */
+void StartIdleResourceCleanupTimers()
+{
+	// We cache the idle session timeout in this backend, because
+	// get_idle_session_timeout_hook() might return different values at a
+	// later time.
+	if (get_idle_session_timeout_hook)
+		IdleSessionTimeout = (*get_idle_session_timeout_hook)();
+	else
+		IdleSessionTimeout = 0;
+
+	int sigalarm_timeout = MinTimeout(get_idle_gang_timeout(),
+	                                  IdleSessionTimeout);
+
+	if (sigalarm_timeout > 0)
+	{
+		if (!enable_sig_alarm(sigalarm_timeout, false))
+			elog(FATAL, "could not set timer for client wait timeout");
```

While not introduced by this patch, I wonder why we consider this a `FATAL`? Seems a big gun to use, but I might be missing something. Not to be changed in this PR however, just thinking out loud.

---

Hmm, well looking into the code of `enable_sigalarm` we see it can only return false if setitimer fails.

```c
	/* If we reach here, okay to set the timer interrupt */
	MemSet(&timeval, 0, sizeof(struct itimerval));
	timeval.it_value.tv_sec = delayms / 1000;
	timeval.it_value.tv_usec = (delayms % 1000) * 1000;
	if (setitimer(ITIMER_REAL, &timeval, NULL))
		return false;
```


setitimer only returns non-zero in these cases (from https://www.systutorials.com/docs/linux/man/2-setitimer/):
```
*RETURN VALUE
On success, zero is returned. On error, -1 is returned, and errno is set appropriately.  

*ERRORS
EFAULT
new_value, old_value, or curr_value is not valid a pointer.
EINVAL
which is not one of ITIMER_REAL, ITIMER_VIRTUAL, or ITIMER_PROF; or (since Linux 2.6.22) one of the tv_usec fields in the structure pointed to by new_value contains a value outside the range 0 to 999999
```

Looks like this is very unlikely to happen, unless something like a negative number is passed in as the timeout or Bad Stuff™ happens in the system, so FATAL is probably OK since we want to fail hard so the bad backend dies (as it doesn't do the proper behavior).

---

Right, we should at least change the error message to include errno and `%m`, but not in this PR.